core/templates/dev/head/services/StateTopAnswersStatsService.js

```diff
@@ -0,0 +1,123 @@
+  'ExplorationContextService', 'ExplorationStatesService',
+  'UrlInterpolationService',
+  function(
+      $http, $injector, AngularNameService, AnswerClassificationService,
+      ExplorationContextService, ExplorationStatesService,
+      UrlInterpolationService) {
+    /**
+     * @typedef {Object} AnswerStats - A record for the statistics of a single
+     * top answer. TODO(brianrodri): Consider placing this into its own domain
+     * object.
+     *
+     * @property {*} answer - Contains the answer in its raw form, directly from
+     *    the backend.
+     * @property {string} answerHtml - Contains the answer as a string which can
+     *    be rendered directly as HTML.
+     * @property {number} frequency
+     * @property {boolean} isAddressed
+     */
+
+    /** @type {Object.<string, AnswerStats[]>} */
+    var stateTopAnswersCache = {};
+
+    /**
+     * TODO(brianrodri): Move this helper function into a proper service which
+     * takes the state's interaction type into account when formatting the HTML.
+     * @param {*} answer
+     * @returns {string}
+     */
+    var convertAnswerToHtml = function(answer) {
+      // Don't render quotes when the answer is a string.
+      return (typeof answer === 'string') ? answer : angular.toJson(answer);
+    };
+
+    /**
+     * @param {string} stateName - target state to have its addressed info
+     * refreshed.
+     */
+    var refreshAddressedInfo = function(stateName) {
+      var explorationId = ExplorationContextService.getExplorationId();
+      var state = ExplorationStatesService.getState(stateName);
+      var interactionRulesService = $injector.get(
+        AngularNameService.getNameOfInteractionRulesService(
+          state.interaction.id));
+
+      stateTopAnswersCache[stateName].forEach(function(answerStats) {
+        answerStats.isAddressed =
+          AnswerClassificationService.isClassifiedExplicitlyOrGoesToNewState(
+            explorationId, stateName, state, answerStats.answer,
+            interactionRulesService);
+      });
+    };
+
+    return {
+      /**
+       * Calls the backend asynchronously to setup the answer statistics of each
+       * state this exploration contains.
+       */
+      init: function() {
+        var explorationId = ExplorationContextService.getExplorationId();
+
+        $http.get(
+          UrlInterpolationService.interpolateUrl(
+            '/createhandler/state_answer_stats/<exploration_id>',
+            {exploration_id: explorationId})
+        ).then(function(response) {
+          stateTopAnswersCache = {};
+          Object.keys(response.data.answers).forEach(function(stateName) {
+            var stateAnswerStatsBackendDicts = response.data.answers[stateName];
+
+            stateTopAnswersCache[stateName] =
+              stateAnswerStatsBackendDicts.map(function(answerFrequencyPair) {
+                return /** @type {AnswerStats} */ {
```

Comment:
Shouldn't this be an AnswerStats domain object?

Reply:
Yep, for now I'm just leaving it as a "documented type" to get the service out first!

Reply:
Oh, I see. Let's not do that -- could you please add the domain object in this PR as well? It should hopefully be straightforward.

I'm generally fine with submitting partial stuff, but I think the intermediate version shouldn't be "incorrect" (which returning a raw dict is). Don't want to propagate this pattern.

Reply:
Done.